### ISSUE TITLE
Consul is continually being used when it is not configured.

### ISSUE THREAD

**Issue by cletisgipson** [ThreeMammals/Ocelot#428]
I am currently using the latest stable release of the product to setup a Proof Of Concept.  Every call I make, throws a Consul.GetRequest error in the logs.   Here is my configuration file.

```
{
  "ReRoutes": [
    {
      "UseServiceDiscovery":  "false",
      "DownstreamPathTemplate": "/api/customers",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 9001
        }
      ],
      "UpstreamPathTemplate": "/customers",
      "UpstreamHttpMethod": [ "Get" ]
    }
  ],
  "GlobalConfiguration": {
    "RequestIdKey": "OcRequestId",
    "BaseUrl":"http://localhost:9000",
  }
}  
```

Windows 10, Visual Studio 2017 with two very basic apis and one gateway project.  The UseServiceDiscovery flag seems to do no good.

**Comment by TomPallister** [comment]
@cletisgipson thanks for your interest in the project! mmm this is very strange might be a bug!

Can you paste your program.cs and startup.cs here as well?

**Comment by TomPallister** [comment]
@cletisgipson also is it possible to get the full stack trace?

**Comment by cletisgipson** [comment]
Program.cs

```
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Hosting;

namespace APIGateway
{
    public class Program
    {
        public static void Main(string[] args)
        {
            BuildWebHost(args).Run();
        }

        public static IWebHost BuildWebHost(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .UseUrls("http://localhost:9000")
                .Build();
    }
}
```

Startup.cs

```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Ocelot.DependencyInjection;
using Ocelot.Middleware;

namespace APIGateway
{
    public class Startup
    {
        public IConfigurationRoot Configuration { get; }


        public Startup(IHostingEnvironment env)
        {
            var builder = new Microsoft.Extensions.Configuration.ConfigurationBuilder();
            builder.SetBasePath(env.ContentRootPath)
                   //add configuration.json  
                   .AddJsonFile("configuration.json", optional: false, reloadOnChange: true)
                   .AddEnvironmentVariables();

            Configuration = builder.Build();
        }

        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {

            services.AddOcelot();

        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public async void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            await app.UseOcelot();
        }
    }
}
```

Stack Trace


C:\Users\cgipson\Source\Repos\OcelotProofOfConcept\APIGateway>dotnet run
Using launch settings from C:\Users\cgipson\Source\Repos\OcelotProofOfConcept\APIGateway\Properties\launchSettings.json...
Hosting environment: Development
Content root path: C:\Users\cgipson\Source\Repos\OcelotProofOfConcept\APIGateway
Now listening on: http://localhost:9000
Application started. Press Ctrl+C to shut down.
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:9000/customers
info: Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware[0]
      requestId: 0HLEOB35TIB99:00000001, previousRequestId: no previous request id, message: EndpointRateLimiting is not enabled for /
info: Ocelot.Authentication.Middleware.AuthenticationMiddleware[0]
      requestId: 0HLEOB35TIB99:00000001, previousRequestId: no previous request id, message: No authentication needed for /customers
info: Ocelot.Authorisation.Middleware.AuthorisationMiddleware[0]
      requestId: 0HLEOB35TIB99:00000001, previousRequestId: no previous request id, message: / route does not require user to be authorised
fail: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware[0]
      requestId: 0HLEOB35TIB99:00000001, previousRequestId: no previous request id, message: Exception caught in global error handler, exception message: Unexpected response, status code BadGateway: {removed}
, exception stack:    at Consul.GetRequest`1.<Execute>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.ServiceDiscovery.Providers.ConsulServiceDiscoveryProvider.<Get>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.LoadBalancers.LoadBalancerFactory.<>c__DisplayClass2_0.<<Get>b__3>d.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.LoadBalancers.NoLoadBalancer.<Lease>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.QueryStrings.Middleware.QueryStringBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Headers.Middleware.HttpRequestHeadersBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Authorisation.Middleware.AuthorisationMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Claims.Middleware.ClaimsBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Authentication.Middleware.AuthenticationMiddleware.<Invoke>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.RequestId.Middleware.ReRouteRequestIdMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Request.Middleware.DownstreamRequestInitialiserMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Headers.Middleware.HttpHeadersTransformationMiddleware.<Invoke>d__6.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Multiplexer.Multiplexer.<Fire>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Multiplexer.Multiplexer.<Multiplex>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware.<Invoke>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Responder.Middleware.ResponderMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Pipeline.MapWhenMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Errors.Middleware.ExceptionHandlerMiddleware.<Invoke>d__4.MoveNext() RequestId: 0HLEOB35TIB99:00000001, exception: Consul.ConsulRequestException: Unexpected response, status code BadGateway:  {removed}
         at Consul.GetRequest`1.<Execute>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.ServiceDiscovery.Providers.ConsulServiceDiscoveryProvider.<Get>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.LoadBalancers.LoadBalancerFactory.<>c__DisplayClass2_0.<<Get>b__3>d.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.LoadBalancers.NoLoadBalancer.<Lease>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.QueryStrings.Middleware.QueryStringBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Headers.Middleware.HttpRequestHeadersBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Authorisation.Middleware.AuthorisationMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Claims.Middleware.ClaimsBuilderMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Authentication.Middleware.AuthenticationMiddleware.<Invoke>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.RequestId.Middleware.ReRouteRequestIdMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Request.Middleware.DownstreamRequestInitialiserMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Headers.Middleware.HttpHeadersTransformationMiddleware.<Invoke>d__6.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Multiplexer.Multiplexer.<Fire>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Multiplexer.Multiplexer.<Multiplex>d__2.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware.<Invoke>d__5.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Responder.Middleware.ResponderMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Middleware.Pipeline.MapWhenMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Ocelot.Errors.Middleware.ExceptionHandlerMiddleware.<Invoke>d__4.MoveNext()
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 552.5746ms 500

**Comment by TomPallister** [comment]
@cletisgipson awesome thanks so much for this. There is one last thing I need to know 😞 Sorry for not asking originally...which version of Ocelot are you currently using?

**Comment by cletisgipson** [comment]
most recent stable release

**Comment by TomPallister** [comment]
@cletisgipson awesome thanks ill look into this asap! No idea why you are seeing this behaviour.

**Comment by TomPallister** [comment]
@cletisgipson OK there are a few things going on here...

If you set your program.cs to the following your config should start working. The way you were setting up your config doesnt seem to work in .net anymore. I assume you are using 2.0+ if you are using the latest version of Ocelot. More info with the MS docs [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-2.1)

```csharp
public static void Main(string[] args)
        {
            new WebHostBuilder()
                    .UseKestrel()
                    .UseUrls("http://localhost:9000")
                    .UseContentRoot(Directory.GetCurrentDirectory())
                    .ConfigureAppConfiguration((hostingContext, config) =>
                    {
                        config
                            .SetBasePath(hostingContext.HostingEnvironment.ContentRootPath)
                            .AddJsonFile("configuration.json")
                            .AddEnvironmentVariables();
                    })
                    .ConfigureServices(s =>
                    {
                        s.AddOcelot();
                    })
                    .ConfigureLogging((hostingContext, logging) =>
                    {
                        //add your logging
                    })
                    .UseIISIntegration()
                    .Configure(app =>
                    {
                        app.UseOcelot().Wait();
                    })
                    .Build()
                    .Run();
        }
```

It is possible that the other style they have in those docs that still uses a Startup.cs will work but you need to pass the config into the startup constructor e.g..

```csharp
public Startup(IHostingEnvironment env, IConfiguration config)
    {
        HostingEnvironment = env;
        Configuration = config;
    }
```

Another thing that you can do is set the `"UseServiceDiscovery":" false"` to `"UseServiceDiscovery": false` as this might be causing it to default to true.

After making these changes I was able to get it working.

Now the reason it was trying to use consul is that Ocelot was not getting your configuration into the system because of the way Startup.cs is working and when there are no ReRoutes in the config it defaults to trying to use Consul to lookup services dynamically. I feel like I might need to add some logging statements to make this clearer so I will keep this ticket open for that.

Hope this helps! Let me know if it doesnt.

**Comment by TomPallister** [comment]
If no re routes and falling back to config log that fact so users can see it
Also not sure if we should be trying to fall back if nothing specified around service discovery in config however you could just be using consul defaults i guess.

Not sure, needs thought.

**Comment by cletisgipson** [comment]
That worked.  Thanks!

**Comment by TomPallister** [comment]
released in nuget package 7.0.7 will be available on nuget soon!
